src/uct/tcp/sockcm/sockcm_ep.c

```diff
@@ -44,9 +47,173 @@ void uct_sockcm_ep_put_sock_id(uct_sockcm_ctx_t *sock_id_ctx)
     ucs_free(sock_id_ctx);
 }
 
+ucs_status_t uct_sockcm_ep_send_client_info(uct_sockcm_iface_t *iface, uct_sockcm_ep_t *ep)
+{
+    ssize_t transfer_len = 0;
+    ssize_t sent_len     = 0;
+    ssize_t sent_off     = 0;
+    uct_md_h dummy_md    = 0;
+    int retries          = 0;
+    ucs_status_t status;
+    unsigned num_resources;
+    uct_tl_device_resource_t *devices;
+    uct_sockcm_conn_param_t conn_param;
+
+    memset(&conn_param, 0, sizeof(uct_sockcm_conn_param_t));
+
+    status = uct_tcp_query_devices(dummy_md, &devices, &num_resources);
+    if (UCS_OK != status || num_resources == 0) {
+        ucs_error("sockcm unable to find a tcp-capable device");
+        status = (status == UCS_OK) ? UCS_ERR_IO_ERROR : status;
+        goto out;
+    }
+    ucs_debug("using device %s at client side\n", devices[0].name);
+
+    /* pack worker address into private data */
+    conn_param.length = ep->pack_cb(ep->pack_cb_arg, devices[0].name,
+                                    (void*)conn_param.private_data);
+    if (conn_param.length < 0) {
+        ucs_error("sockcm client (iface=%p, ep = %p) failed to fill "
+                  "private data. status: %s",
+                  iface, ep, ucs_status_string(conn_param.length));
+        status = UCS_ERR_IO_ERROR;
+        goto out;
+    }
+    ucs_assert(conn_param.length <= UCT_SOCKCM_PRIV_DATA_LEN);
+    transfer_len = sizeof(uct_sockcm_conn_param_t) - UCT_SOCKCM_PRIV_DATA_LEN;
+    transfer_len += conn_param.length;
+
+    do {
+        sent_len = send(ep->sock_id_ctx->sock_id, (char *) &conn_param + sent_off,
+                        transfer_len - sent_off, 0);
+        ucs_debug("sockcm_client: send_len = %d bytes %m", (int) sent_len);
+        if ((-1 == sent_len) && (errno != EAGAIN && errno != EWOULDBLOCK)) {
+            ucs_error("unable to send client info %s", strerror(errno));
+            status = UCS_ERR_IO_ERROR;
+            ep->conn_state = UCT_SOCKCM_EP_CONN_STATE_CLOSED;
+            goto out;
+        }
+        sent_off += sent_len;
+        retries++;
+    } while (sent_off != transfer_len && retries < 256);
+
+    if (sent_off == transfer_len) {
+        status = UCS_OK;
+        ep->conn_state = UCT_SOCKCM_EP_CONN_STATE_INFO_SENT;
+    } else {
+        status = UCS_ERR_IO_ERROR;
+        ep->conn_state = UCT_SOCKCM_EP_CONN_STATE_CLOSED;
+    }
+
+out:
+    ucs_free(devices);
+    return status;
+}
+
 static void uct_sockcm_ep_event_handler(int fd, void *arg)
 {
-    ucs_debug("not implemented");
+    uct_sockcm_iface_t *iface = NULL;
+    ssize_t recv_len          = 0;
+    uct_sockcm_ep_t *ep       = (uct_sockcm_ep_t *) arg;
+    char conn_status;
+    int optval;
+    socklen_t optlen;
+    ucs_status_t status;
+
+    iface = ucs_derived_of(ep->super.super.iface, uct_sockcm_iface_t);
```

Comment:
can initialize in place (after `ep`), or do not need to init to `NULL` on line 115